concent_api/middleman/tests/test_queue_operations.py

```diff
@@ -0,0 +1,509 @@
+from middleman_protocol.message import GolemMessageFrame
+from middleman_protocol.registry import create_middleman_protocol_message
+from middleman_protocol.stream import append_frame_separator
+from middleman_protocol.stream import escape_encode_raw_message
+from middleman_protocol.tests.utils import prepare_mocked_reader
+from middleman_protocol.tests.utils import prepare_mocked_writer
+
+(CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY) = generate_ecc_key_pair()
+(SIGNING_SERVICE_PRIVATE_KEY, SIGNING_SERVICE_PUBLIC_KEY) = generate_ecc_key_pair()
+FROZEN_DATE_AND_TIME = "2012-01-14 12:00:01"
+FROZEN_TIMESTAMP = datetime.datetime.strptime(FROZEN_DATE_AND_TIME, "%Y-%m-%d %H:%M:%S").timestamp()
+
+
+async def get_item(queue):
+    await sleep(0.000001)
+    item = await queue.get()
+    queue.task_done()
+    return item
+
+
+def _get_mocked_reader(message, request_id, sign_as, **kwargs):
+    protocol_message = create_middleman_protocol_message(PayloadType.GOLEM_MESSAGE, message, request_id)
+    data_to_send = append_frame_separator(
+        escape_encode_raw_message(
+            protocol_message.serialize(sign_as)
+        )
+    )
+    mocked_reader = prepare_mocked_reader(data_to_send, **kwargs)
+    return mocked_reader
+
+
+@freeze_time(FROZEN_DATE_AND_TIME)
+class TestRequestProducer:
+    golem_message = None
+    connection_id = None
+    request_id = None
+    queue = None
+    response_queue = None
+
+    @pytest.fixture(autouse=True)
+    def setUp(self, event_loop):
+        self.golem_message = Ping()
+        self.connection_id = 1
+        self.request_id = 777
+        self.queue = Queue(loop=event_loop)
+        self.response_queue = Queue(loop=event_loop)
+
+    @pytest.mark.asyncio
+    async def test_that_when_valid_frame_is_received_a_new_item_is_added_to_the_queue(self, event_loop):
+        with override_settings(
+            CONCENT_PRIVATE_KEY=CONCENT_PRIVATE_KEY,
+            CONCENT_PUBLIC_KEY=CONCENT_PUBLIC_KEY,
+        ):
+            mocked_reader = _get_mocked_reader(self.golem_message, self.request_id, settings.CONCENT_PRIVATE_KEY)
+
+            producer_task = event_loop.create_task(
+                request_producer(self.queue, self.response_queue, mocked_reader, self.connection_id)
+            )
+            item = await get_item(self.queue)
+            producer_task.cancel()
+
+            assert_that(item.connection_id).is_equal_to(self.connection_id)
+            assert_that(item.concent_request_id).is_equal_to(self.request_id)
+            assert_that(item.message).is_equal_to(self.golem_message)
+            assert_that(item.timestamp).is_equal_to(FROZEN_TIMESTAMP)
+
+    @pytest.mark.asyncio
+    async def test_that_when_client_terminates_connection_coroutine_ends(self, event_loop):
+        with override_settings(
+            CONCENT_PRIVATE_KEY=CONCENT_PRIVATE_KEY,
+            CONCENT_PUBLIC_KEY=CONCENT_PUBLIC_KEY,
+        ):
+            some_bytes = b"some bytes"
```

Comment:
Please use `os.urandom(x)`

Reply:
why? it does not matter and unit tests should be non-deterministic

Reply:
I agree with @dybi. Also note that `os.urandom()` is an overkill in tests. It returns cryptographically strong random numbers straight from `/dev/urandom` and cannot be seeded. That's great when you're e.g. generating a production key. But in tests you want numbers that just look random and you want to be able to repeat exactly the same sequence when you run the test again.